src/Ed25519.hs

```diff
@@ -0,0 +1,269 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE MultiWayIf #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+
+-- Based on https://ed25519.cr.yp.to/python/ed25519.py
+module Ed25519 where
+
+import Control.Monad (unless)
+import Crypto.Sign.Ed25519 (PublicKey (..), SecretKey (..), Signature (..))
+import Crypto.Sign.Ed25519 qualified as Reference
+import Data.ByteString qualified as BS
+import Data.ByteString.Char8 qualified as S8
+import Data.Char (intToDigit)
+import Data.Word qualified as P
+import Debug.Trace (traceM)
+import Debug.Trace qualified as Debug
+import GHC.ByteOrder (ByteOrder (..))
+import Numeric (showHex, showIntAtBase)
+import PlutusTx.Builtins
+  ( BuiltinByteString,
+    Integer,
+    appendByteString,
+    byteStringToInteger,
+    divideInteger,
+    equalsInteger,
+    integerToByteString,
+    lengthOfByteString,
+    modInteger,
+    multiplyInteger,
+    readBit,
+    sliceByteString,
+    toBuiltin,
+    writeBits,
+  )
+import PlutusTx.Builtins.Internal (BuiltinBool, BuiltinByteString (..), BuiltinList (..), BuiltinPair (..))
+import PlutusTx.List (replicate)
+import PlutusTx.Prelude
+  ( AdditiveGroup ((-)),
+    AdditiveSemigroup ((+)),
+    Bool (..),
+    Eq ((==)),
+    MultiplicativeSemigroup ((*)),
+    Semigroup ((<>)),
+    divMod,
+    fst,
+    not,
+    odd,
+    otherwise,
+    product,
+    snd,
+    ($),
+    (&&),
+    (.),
+    (/=),
+  )
+import SHA512V2 (doTrace, prettify, sha512)
+import Prelude qualified as P
+
+showHex' :: BuiltinByteString -> [P.Char]
+showHex' (BuiltinByteString bi) = P.concatMap (\x -> pad $ showHex x []) (BS.unpack bi)
+  where
+    pad [x] = ['0', x]
+    pad s = s
+
+-- No checks for size, though we should be doing them.
+checkValid ::
+  BuiltinByteString ->
+  BuiltinByteString ->
+  BuiltinByteString ->
+  Bool
+checkValid sig message pubKey =
+  let sliceR = (sliceByteString 0 32 sig)
+      !r = decodePoint (sliceByteString 0 32 sig)
+      !a = decodePoint pubKey
+      !s = decodeInt (sliceByteString 32 32 sig)
+      hintArg = (encodePoint r <> pubKey <> message)
+      !h = hint hintArg
+      msg =
+        prettify
+          [ "checkValid",
+            "sig len: " <> P.show (lengthOfByteString sig),
+            "nmsg len: " <> P.show (lengthOfByteString message),
+            "pk len: " <> P.show (lengthOfByteString pubKey),
+            "hint " <> showHex' hintArg <> " = " <> P.show h,
+            "sig: " <> P.show (showHex' sig),
+            "msg: " <> P.show (showHex' message),
+            "pubKey: " <> P.show (showHex' pubKey),
+            "r: " <> P.show r,
+            "a: " <> P.show a,
+            "s: " <> P.show s,
+            "sliceR: " <> showHex' sliceR
+          ]
+   in doTrace msg $ eqPoint (scalarMult bPoint s) (edwards r (scalarMult a h))
```

Comment:
We probably don't want to leave in the tracing here: onchain, this would be _extremely_ expensive. Also, `P.show` won't compile anyway, nor will `showHex'` as-written, since they use things from the Haskell prelude.